### ISSUE TITLE
"cannot start a transaction within a transaction" with sqflite backend on iOS

### ISSUE THREAD

**Issue by Mike278** [simolus3/moor#2470]
We're beta testing drift 2.8.1 using the drift_sqflite 2.0.0 backend, and we're experiencing the UI waiting for queries to complete and the following issue popped up on Crashlytics:
```
DatabaseException(Error Domain=FMDatabase Code=1 "cannot start a transaction within a transaction" UserInfo={NSLocalizedDescription=cannot start a transaction within a transaction}) sql 'BEGIN IMMEDIATE' args []
```

This only happens on iOS devices (Android works) and only using drift_sqflite (switching to the dart:ffi sqlite3 backend fixes it on iOS). 

Read and write queries seem to work fine when tapping around the app. This only happens when the app is syncing, during which:
- queries are running in response to concurrent network requests (i.e. we're relying on drift to synchronize concurrent reads/writes)
- batches are persisted inside transactions (using the `batchInsertAll` function below)
- the function passed to `transaction` may throw
- the function passed to `transaction` may try/catch

Unfortunately we don't have a usable stack trace atm and we're in the process of improving logging here to get more details. Still trying to narrow down exact steps to reproduce (don't have a Mac to debug with atm..), but FWIW when trying to reproduce this on Android we see logs that indicate nested `SAVEPOINT`s/`RELEASE`s up to 3 levels deep - which I think is expected - but just wanted to mention it.

<details>
<summary>batchInsertAll</summary>

```dart
  /// By default the [QueryEngine.batch] api throws on the first error, and if uncaught, also
  /// rolls back the transaction, therefore reverting all other changes made by the batch.
  ///
  /// This function is a wrapper around the [QueryEngine.batch] api that
  /// enables per-record error reporting and avoids transaction rollbacks.
  ///
  /// Each time the batch encounters an error [onError] is invoked,
  /// and the batch is restarted from 1-after the failed record.
  Future<void> batchInsertAll<T>({
    required Iterable<T> items,
    required void Function(Batch, T) insert,
    required void Function(T, Object) onError,
  }) {
    return transaction(() async {
      final offset = await totalChanges();
      int index = 0, failed = 0;
      final list = [...items];
      while (index < items.length) {
        try {
          return await batch((batch) {
            for (int i = index; i < items.length; i++) {
              insert(batch, list[i]);
            }
          });
        }
        on Object catch (error) {
          final saved = index - ++failed;
          final inserted = await totalChanges() - (offset + saved);
          index += inserted;
          final failedItem = list[index-1];
          onError(failedItem, error);
        }
      }
    });
  }
```
</details>

**Comment by Mike278** [comment]
Looks like it might be related to the nested transactions feature added in 2.0? It's tested in 48830b57aa7611ceab2449e69cf4bbfeef398047 though... hmm, still looking.

**Comment by simolus3** [comment]
Weirdly, test passes on Android, but fails on iOS:

```dart
  test('handles failing commits', () async {
    final executor = SqfliteQueryExecutor(path: ':memory:');
    final database = EmptyDb(executor);
    addTearDown(database.close);

    await database.customStatement('PRAGMA foreign_keys = ON;');
    await database.customStatement('CREATE TABLE x (foo INTEGER PRIMARY KEY);');
    await database.customStatement('CREATE TABLE y (foo INTEGER PRIMARY KEY '
        'REFERENCES x (foo) DEFERRABLE INITIALLY DEFERRED);');

    await expectLater(
      database.transaction(() async {
        await database.customStatement('INSERT INTO y VALUES (2);');
      }),
      throwsA(isA<CouldNotRollBackException>()),
    );
  });
```

On Android, we receive a `CouldNotRollbackException` because the failing `COMMIT` due to the foreign key constraint also rolls back the exception implicitly. On iOS and my distro's sqlite3 library, the `ROLLBACK` works. Weirdly, the rollback doesn't work in the sqlite3 CLI which uses the same shared library. Maybe there's some pragma going on here.

I'm now wondering if we might hit a scenario where Android rolls back transactions for us but we're not doing that? But that should also be the case in the `drift/native.dart` backend then.

**Comment by simolus3** [comment]
> FWIW when trying to reproduce this on Android we see logs that indicate nested `SAVEPOINT`s/`RELEASE`s up to 3 levels deep - which I think is expected - but just wanted to mention it.

Hold on, `batch` inside a transaction should _not_ start its own nested savepoint scope (arguably an outdated design choice now that nested transactions are supported, but it's mentioned in its dartdocs). Are you calling `batchInsertAll` in a double nested transaction?

**Comment by Mike278** [comment]
> Are you calling batchInsertAll in a double nested transaction?

Finally got some better visibility and it looks like that's exactly what's happening - and it's actually a *quadruple* nested transaction! 


<details>
<summary>Somewhat formatted relevant snippet of logs</summary>

```
2023-06-08 21:16:26.891031: Drift: Sent BEGIN TRANSACTION with args []
2023-06-08 21:16:26.891726: Drift: Sent select * from [snip]
2023-06-08 21:16:26.892360: Drift: Sent select [snip]
2023-06-08 21:16:26.893121: Drift: Sent SAVEPOINT s0 with args []
2023-06-08 21:16:26.893359: Drift: Sent SELECT value FROM trigger_settings WHERE id = 'change_tracking_enabled' with args []
2023-06-08 21:16:26.893583: Drift: Sent UPDATE trigger_settings SET value = 0 WHERE id = 'change_tracking_enabled' with args []
2023-06-08 21:16:26.894801: Drift: Sent SAVEPOINT s1 with args []
2023-06-08 21:16:26.894944: Drift: Sent SELECT value FROM trigger_settings WHERE id = 'change_tracking_enabled' with args []
2023-06-08 21:16:26.895124: Drift: Sent UPDATE trigger_settings SET value = 0 WHERE id = 'change_tracking_enabled' with args []
2023-06-08 21:16:26.895244: Drift: Sent SAVEPOINT s2 with args []
2023-06-08 21:16:26.895397: Drift: Sent select total_changes() with args []
2023-06-08 21:16:26.896054: Drift: Executing BatchedStatements(
  [INSERT OR REPLACE INTO "some_table" ("some_id") VALUES (?)], 
  [
    ArgumentsForBatchedStatement(0, [88c87d75-6215-4655-a2ef-5e2265cf5ebc]), 
    ArgumentsForBatchedStatement(0, [383acb40-b96f-48ae-9a68-22864d7f459c]), 
    ArgumentsForBatchedStatement(0, [a61422bb-dd76-432d-9e5d-f4df24873989]),
  ]
) 
2023-06-08 21:16:26.896399: Drift: Sent select total_changes() with args []
2023-06-08 21:16:26.896843: Drift: Executing BatchedStatements(
  [INSERT OR REPLACE INTO "some_table" ("some_id") VALUES (?)], 
  [
    ArgumentsForBatchedStatement(0, [383acb40-b96f-48ae-9a68-22864d7f459c]), 
    ArgumentsForBatchedStatement(0, [a61422bb-dd76-432d-9e5d-f4df24873989]),
  ]
) 
2023-06-08 21:16:26.897081: Drift: Sent select total_changes() with args []
2023-06-08 21:16:26.897296: Drift: Executing BatchedStatements(
  [INSERT OR REPLACE INTO "some_table" ("some_id") VALUES (?)], 
  [
    ArgumentsForBatchedStatement(0, [a61422bb-dd76-432d-9e5d-f4df24873989]),
  ]
) 
2023-06-08 21:16:26.897554: Drift: Sent select total_changes() with args []
2023-06-08 21:16:26.897735: Drift: Sent RELEASE s2 with args []
2023-06-08 21:16:26.897898: Drift: Sent UPDATE trigger_settings SET value = 0 WHERE id = 'change_tracking_enabled' with args []
2023-06-08 21:16:26.898038: Drift: Sent RELEASE s1 with args []
2023-06-08 21:16:26.898205: Drift: Sent UPDATE trigger_settings SET value = 1 WHERE id = 'change_tracking_enabled' with args []
2023-06-08 21:16:26.898358: Drift: Sent RELEASE s0 with args []
2023-06-08 21:16:26.898553: Drift: Sent COMMIT TRANSACTION with args []
2023-06-08 21:16:26.899247: SyncAPI: Failed to save [someId=88c87d75-6215-4655-a2ef-5e2265cf5ebc] DatabaseException(Error Domain=FMDatabase Code=1 "cannot start a transaction within a transaction" UserInfo={NSLocalizedDescription=cannot start a transaction within a transaction}) sql 'BEGIN IMMEDIATE' args []
2023-06-08 21:16:26.899765: SyncAPI: Failed to save [someId=383acb40-b96f-48ae-9a68-22864d7f459c] DatabaseException(Error Domain=FMDatabase Code=1 "cannot start a transaction within a transaction" UserInfo={NSLocalizedDescription=cannot start a transaction within a transaction}) sql 'BEGIN IMMEDIATE' args []
2023-06-08 21:16:26.900270: SyncAPI: Failed to save [someId=a61422bb-dd76-432d-9e5d-f4df24873989] DatabaseException(Error Domain=FMDatabase Code=1 "cannot start a transaction within a transaction" UserInfo={NSLocalizedDescription=cannot start a transaction within a transaction}) sql 'BEGIN IMMEDIATE' args []

```
</details>

- The top level `BEGIN TRANSACTION` is from a function that's called when a network request completes
- `SAVEPOINT s0` and `SAVEPOINT s1` are calls to a function that disables some insert/update triggers that track local changes performed by the user - roughly:
```dart
Future<T> withoutChangeTracking<T>(Future<T> Function() fn) =>
  db.transaction(() async {
    bool previous = await changeTrackingEnabled();
    try {
      await enableChangeTracking(false);
      return await fn();
    }
    finally {
      await changeTrackingEnabled(previous);
    }
  });
```
- (Note: the creation of `s1` is redundant)
- `SAVEPOINT s2` is the transaction that `batchInsertAll` starts
- You can see `batchInsertAll` restarting the batch 3 times as each of the 3 items fail to insert
- After all the savepoints are released and the top level transaction is committed, the errors encountered during `batchInsertAll` are logged

Interestingly, the error says the sql is `BEGIN IMMEDIATE`, but there aren't any logs of Drift sending that statement. I couldn't find any references to `BEGIN IMMEDIATE` in Drift, but I did find one in sqflite [here](https://github.com/tekartik/sqflite/blob/70382fc1c996d77af6d79e7f555470b2aa2a0ef3/sqflite_common/lib/src/database_mixin.dart#L356). So seems like sqflite's batch is trying to start a transaction?

**Comment by simolus3** [comment]
I guess that's what's happening here. We're already using the variant of batches that doesn't use transactions though (by calling `batch.apply()` instead of `batch.commit()` in `_SqfliteDelegate.runBatched`). I'll see if I can reproduce this, perhaps it's a regression in sqflite or in the way we start batches in drift.

**Comment by simolus3** [comment]
Oh yikes, I forgot to upload that fix to pub!

Can you reproduce this still if you're using `drift_sqflite` with a git dependency override to this repository? There might be something else going on, but I guess that should fix it. If it does, I'll release it right away.

**Comment by Mike278** [comment]
Ah nice find! That might also explain 
> On iOS and my distro's sqlite3 library, the ROLLBACK works. Weirdly, the rollback doesn't work in the sqlite3 CLI which uses the same shared library

I'll try it out and let you know 👍

**Comment by Mike278** [comment]
Yep that fixes it! 🎉 Easiest bugfix ever :P

**Comment by simolus3** [comment]
Thanks for the confirmation, fixed in `drift_sqflite: ^2.0.1`.
